### PR TITLE
MAINT: Added pickle protocol methods into zipline.

### PR DESCRIPTION
**Pull request by CaptainKanuk**
For review only. Tests will be added soon.

Added pickle support to many zipline methods. This will enable
them to be serialized.

### CODE REVIEW

**Comment by dalejung** [comment]
@CaptainKanuk just wanted to confirm that it's good for review now?

**Comment by CaptainKanuk** [comment]
Yes, I'll just be adding unit tests to this.

**Comment by CaptainKanuk** [comment]
I ended up adding code to period, tracker and position for consistency. It's not much of a change code-lines wise, so it should be okay to rebase over.

File: zipline/utils/serialization_utils.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    def __setstate__(self, state):
+        """
+        Many objects require only this code.
+        """
+        self.__dict__.update(state)
+
+    # =====================================================
+    # These are helper methods for some problem data types.
+    # =====================================================
+
```
Comment (dalejung) [inline zipline/utils/serialization_utils.py]: remove these helpers and handle them in `__setstate__`

Comment (dalejung) [inline zipline/utils/serialization_utils.py]: Something like this: https://github.com/quantopian/zipline/pull/505/files#diff-f3d976ceea2b1ea116269ac8b7e9ccc8R428

These objects needs to be internally consistent. These objects won't inflate without the internal stuff that knows what those special types mean.

Comment (CaptainKanuk) [inline zipline/utils/serialization_utils.py]: Yeah I see the issue.

File: zipline/utils/serialization_utils.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Label for the serialization version field in the state returned by
+# __getstate__.
+VERSION_LABEL = '_stateversion_'
+
+
+class SerializeableZiplineObject(object):
```
Comment (dalejung) [inline zipline/utils/serialization_utils.py]: Why do we need to have this class at all? The getstate method is a perfectly good utility func. I think it was basic_get_state before. There really not a single subclass that doesn't override those methods.

File: zipline/finance/slippage.py [hunk 3]
```diff
@@ -128,6 +132,25 @@ def to_dict(self):
...
+
+        return state_dict
+
+    def __setstate__(self, state):
+
+        OLDEST_SUPPORTED_STATE = 1
+        version = state.pop(VERSION_LABEL)
+
+        if version < OLDEST_SUPPORTED_STATE:
+            raise BaseException("Transaction saved state is too old.")
```
Comment (dalejung) [inline zipline/finance/slippage.py]: can we refactor this logic out into a check_versino function along with a custom error? I don't mind the BaseException but joe will yell at me if I don't ask for a custom exception :/

**Comment by dalejung** [comment]
More of a general comment about the state we are saving. Since we are versioning state, we want to keep the state as concise as possible. Values that are temporal like `_position_store` should not be stored because they don't really reflect the state of the object. 

For the most part, it is better to very explicit in what state you save so that you don't grab hanger ons. While they won't break the serialization, they will change the state schema and thus require different versions if/when we clean up or refactor.

type specific data like defaultdict and what not are example of extraneous information. The object needs to know position values from state, it does not need to know how these positions were stored and where on the object they were stored. It already knows this. The `__dict__` as state thing is more a quirk and dirty hack for pickling, but too brittle for when you care about versioning.

I bring up the defaultdict specifically because I am going to remove them game of thrones style. They are quite literally the devil.

**Comment by jbredeche** [comment]
Inviting the defaultdict to the red wedding?

File: zipline/utils/serialization_utils.py (hunk 1 above)
Comment (CaptainKanuk) [inline zipline/utils/serialization_utils.py]: Fixed

**Comment by dalejung** [comment]
https://github.com/quantopian/zipline/pull/525 has tests to grab the failures. 

```
======================================================================
FAIL: test_object_serialization_18_Portfolio (tests.test_serialization.SerializationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/datacaliber/miniconda/envs/qexec/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/Users/datacaliber/work/quantopian/zipline/tests/test_serialization.py", line 123, in test_object_serialization
    self.assertEqual(len(bad_dicts), 0, msg)
AssertionError: Only support bare dicts. type(positions) == <class 'zipline.protocol.Positions'>

======================================================================
FAIL: test_object_serialization_5_PerformancePeriod (tests.test_serialization.SerializationTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/datacaliber/miniconda/envs/qexec/lib/python2.7/site-packages/nose_parameterized/parameterized.py", line 233, in <lambda>
    standalone_func = lambda *a: func(*(a + p.args), **p.kwargs)
  File "/Users/datacaliber/work/quantopian/zipline/tests/test_serialization.py", line 123, in test_object_serialization
    self.assertEqual(len(bad_dicts), 0, msg)
AssertionError: Only support bare dicts. type(orders_by_id) == <type 'cyordereddict._cyordereddict.OrderedDict'>

----------------------------------------------------------------------
```

Are the two failures on my machine.

**Comment by dalejung** [comment]
btw, i dunno if you can just merge this PR into your branch. Give it a shot. I'm curious.

**Comment by CaptainKanuk** [comment]
What do you mean by merge this PR into my branch?

**Comment by dalejung** [comment]
#525 has tests to merge. I didn't have commit access. I dunno if you can just merge it into a branch.

**Comment by dalejung** [comment]
so it looks like if I create a PR off of a branch, it'll merge into the branch and not master. whee

**Comment by CaptainKanuk** [comment]
@dalejung  I'm honestly not sure how to deal with the case of your PR being pulled off my branch.

**Comment by dalejung** [comment]
lol neither do i. If you want to just copy the diff over that’ll work. Is it possible to click the PR interface and merge it? and then just pull from github? 


—

**Comment by CaptainKanuk** [comment]
@dalejung I see what you're doing. I'll copy over the diff and incorporate it into my newest code.

**Comment by dalejung** [comment]
cool cool. I think that should be enough to catch the pickle stuff earlier.


—

**Comment by CaptainKanuk** [comment]
@dalejung That test helper stuff you wrote is on point.

**Comment by llllllllll** [comment]
do you think that it would make sense to archive and compress all of the testing save states?  These files make the repo considerably larger and are only for testing; maybe we should rewrite the tests to unzip and unarchive the tests so that they have a smaller presence on disk

**Comment by CaptainKanuk** [comment]
I think that's a good idea. I'll put that on my TODO list.

**Pull request merged**